services/scribe/backfill/fetcher.go

```diff
@@ -0,0 +1,210 @@
+		iterator:        util.NewChunkIterator(startBlock, endBlock, chunkSize, true),
+		startBlock:      startBlock,
+		endBlock:        endBlock,
+		logs:            make(chan []types.Log, bufferSize),
+		backend:         backend,
+		contractAddress: address,
+		doneChan:        make(chan bool),
+		chainConfig:     chainConfig,
+	}
+}
+
+// closeOnDone closes the done channel when the process is finished.
+func (f *RangeFilter) closeOnDone() {
+	f.doneChan <- true
+}
+
+// GetChunkArr gets the appropriate amount of block chunks (getLogs ranges).
+func (f *RangeFilter) GetChunkArr() (chunkArr []*util.Chunk) {
+	for i := uint64(0); i < f.chainConfig.GetLogsBatchAmount; i++ {
+		chunk := f.iterator.NextChunk()
+		if chunk == nil {
+			return chunkArr
+		}
+		chunkArr = append(chunkArr, chunk)
+
+		// Stop appending chunks if the max height of the current chunk exceeds the concurrency threshold
+		if chunk.EndBlock.Uint64() > f.endBlock.Uint64()-f.chainConfig.ConcurrencyThreshold {
+			return chunkArr
+		}
+	}
+	return chunkArr
+}
+
+// Start starts the log fetching process. If the context is canceled, logs will stop being filtered.
+// 1. Within an infinite for loop, chunks of getLogs blocks are constructed and used to get logs. This flow is paused
+// when the logs channel's buffer of 15 is reached.
+// 2. Each time the logs are received, a wait group is used to ensure that there is no race condition
+// where channels could be closed before a log could be saved.
+// 3. When the range to get logs is completed (GetChunkArr returns a zero array), the wait group is used to ensure
+// that all logs are added to the logs channel before returning and terminating the function.
+// 4. Completing the Start function triggers the closeOnDone function, which sends a boolean in the done channel
+// that signals that the fetcher has completed. The consumer of these logs then performs a drain to fully empty the logs
+// channel. See contract.go to learn more how the logs from this file are consumed.
+func (f *RangeFilter) Start(ctx context.Context) error {
+	var wg sync.WaitGroup
+
+	defer f.closeOnDone()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() != nil {
+				LogEvent(WarnLevel, "could not finish filtering range", LogData{"ca": f.contractAddress, "sh": f.startBlock.String(), "eh": f.endBlock.String(), "cid": &f.chainConfig.ChainID})
+				return fmt.Errorf("could not finish filtering range: %w", ctx.Err())
+			}
+
+			return nil
+		default:
+			chunks := f.GetChunkArr()
+
+			if len(chunks) == 0 {
+				wg.Wait()
+				return nil
+			}
+			logs, err := f.FilterLogs(ctx, chunks)
+			if err != nil {
+				return fmt.Errorf("could not filter logs: %w", err)
+			}
+
+			wg.Add(1)
+			go func(logs []types.Log) {
+				defer wg.Done()
+				f.logs <- logs
+			}(logs)
+			LogEvent(InfoLevel, "Contract backfill chunk completed", LogData{"ca": f.contractAddress, "sh": chunks[0].MinBlock(), "eh": chunks[0].MaxBlock(), "cid": &f.chainConfig.ChainID})
```

Comment:
wtf are ca, sh, eh, etc

Reply:
its for mapping attributes for logging (see logger)

obviously, this is overboard, its in the description of this pr to get do this with error codes (as a stretch goal)